datadog/resource_datadog_security_monitoring_rule.go

```diff
@@ -317,35 +363,78 @@ func resourceDatadogSecurityMonitoringRuleCreate(ctx context.Context, d *schema.
 		return diag.FromErr(err)
 	}
 
-	d.SetId(response.SecurityMonitoringStandardRuleResponse.GetId())
+	if response.SecurityMonitoringStandardRuleResponse != nil {
+		d.SetId(response.SecurityMonitoringStandardRuleResponse.GetId())
+	} else if response.SecurityMonitoringSignalRuleResponse != nil {
+		d.SetId(response.SecurityMonitoringSignalRuleResponse.GetId())
+	} else {
+		return diag.FromErr(fmt.Errorf("SecurityMonitoringStandardRuleResponse and SecurityMonitoringSignalRuleResponse are both empty"))
+	}
+
+	return nil
+}
+
+func isSignalCorrelationSchema(d *schema.ResourceData) bool {
+	if v, ok := d.GetOk("type"); ok {
+		_, err := datadogV2.NewSecurityMonitoringSignalRuleTypeFromValue(v.(string))
+		return err == nil
+	}
+	return false
+}
 
+func checkQueryConsistency(d *schema.ResourceData) error {
+	query := d.Get("query").([]interface{})
+	signalQuery := d.Get("signal_query").([]interface{})
+	if len(query) > 0 && len(signalQuery) > 0 {
+		return fmt.Errorf("query list and signal query list cannot be both populated")
+	}
+	isSignalCorrelation := isSignalCorrelationSchema(d)
+	if !isSignalCorrelation && len(signalQuery) > 0 {
+		return fmt.Errorf("signal query list should not be populated for this rule type")
+	}
+	if isSignalCorrelation && len(query) > 0 {
+		return fmt.Errorf("query list should not be populated for this rule type")
+	}
 	return nil
 }
 
 func buildCreatePayload(d *schema.ResourceData) (datadogV2.SecurityMonitoringRuleCreatePayload, error) {
+
+	if err := checkQueryConsistency(d); err != nil {
+		return datadogV2.SecurityMonitoringRuleCreatePayload{}, err
+	}
+	if isSignalCorrelationSchema(d) {
+		payload, err := buildCreateSignalPayload(d)
+		return datadogV2.SecurityMonitoringSignalRuleCreatePayloadAsSecurityMonitoringRuleCreatePayload(&payload), err
+	}
+	payload, err := buildCreateStandardPayload(d)
+	return datadogV2.SecurityMonitoringStandardRuleCreatePayloadAsSecurityMonitoringRuleCreatePayload(&payload), err
+}
+
+func buildCreateStandardPayload(d *schema.ResourceData) (datadogV2.SecurityMonitoringStandardRuleCreatePayload, error) {
 	payload := datadogV2.SecurityMonitoringStandardRuleCreatePayload{}
 	payload.Cases = buildCreatePayloadCases(d)
```

Comment:
Is there a reason to change all the other fields to `payload.SetXxx` but not `Cases`?

Reply:
I missed it when doing the update. I will fix that.